### DISCUSSION TITLE
How to actually do link prediction

### DISCUSSION THREAD

**Discussion by sinaziaee** [pyg-team/pytorch_geometric#4012] [Q&A] [answered]
Hello everyone and thanks a lot for both the creation of pyg and the great documentation.
I've been following the examples on the examples directory and I came a cross link prediction at [this link](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py). It was a great scenario for understanding this concept but I was actually wondering where the real prediction based on the available data is happening. We split the data into train val and test to compute the loss and accuracy of val and test but where does it actually predict new edges based on the available data. The part that negative sampling is happening is where we want to predict does neg_edges based on our pos_edges but what about when we want to use all the edges as pos_edge and predict new edges. Sorry if it is a simple question but I'm a rookie at gnn and pyg and I would really appreciate all your help and if it was possible to attach a sample code or suggest a link for this purpose that would be really helpful. Thanks guys.

**Comment by rusty1s** [comment, marked as answer]
The prediction of new edges happens inside the decoder part of the model. That is, we use the existing edges in the graph to encode node embeddings using a GNN (encoder part), and then use these node embeddings to find new links (decoder part). You can best see this [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py#L96-L97), where `decode_all` produces a new adjacency matrix. Let me know if this clarifies your issues.

**Reply by sinaziaee** [reply]
So you basically mean that these two lines of codes:

z = model.encode(test_data.x, test_data.edge_index)
final_edge_index = model.decode_all(z)

are predicting new links based on all of our available data ?
Am I correct  ?
So that "final_edge_index" includes both available edges and new predicted edges ?

**Reply by rusty1s** [reply]
The `final_edge_index` contains all the predicted edges for all pairs of nodes.

**Reply by sinaziaee** [reply]
Thank you Matthias. You are really helpful. I've been following your work for a while, specially on Stanford workshop last year. Thank you for all of your help.

**Reply by rusty1s** [reply]
Thanks. Let me know if you have any further questions.

**Reply by sinaziaee** [reply]
Hello again matthias. Sorry for asking again. I've been playing with the "link_pred.py" for some time, and as you said the "final_edge_index" contains all the predicted edges for all pairs of nodes. the problem is the number of these edges are so many. I mean based on the Cora graph we have something around 10556 edges and after running the training phase and using the decode_all method we get something around 3296746 edges with prod_adj > 0 which is so many, so it must not be correct :) . How's that possible ?
Even with this amount of edges it does not contain all the edges that were inside the graph itself(some of the 10556 edges are missing).
Isn't it supposed to at least have all the graph edges at the first place and then predict new edges ? sorry if it's a silly question, but I'm really stuck here.
Even if I change the code so that it produces less edges with higher prod_adj it predicts even less number of original edges. I thought maybe the original edges might have higher prod_adj values so they must not be bothered with setting the threshold higher like returning edges with prod_adj > 5.  What do I have to do to get all the original edges correctly and then some new edges in decode_all function or somewhere else.
The example source code that I'm talking about is [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py#L96-L97).
Thanks

**Reply by sinaziaee** [reply]
Also one another problem. why prob_adj is not sth between 0 and 1 ? isn't this a probability ?

**Reply by rusty1s** [reply]
It is not guaranteed that all existing edges are also part of the predicted ones. If you want to guarantee that, you would have to merge them, e.g.:
```python
edge_index = torch.cat([orig_edge_index, pred_edge_index], dim=1)
edge_index, _ = coalesce(edge_index)  # Remove duplicates
```
Increasing the threshold on `prob_adj` definitely helps to maintain only the edges that should be connected with higher probability. 

**Edit:** Note that the decoder part of the network is untrained/heavily regularized, so it is expected that not all training edges can be recovered.

**Reply by sinaziaee** [reply]
Oh that's right. Thank you very much Matthias.
Do you have any example on how to have a trained decoder for link prediction ?
Just like the encoding part that we can train in the training loop.

**Reply by rusty1s** [reply]
The most simple way to achieve this is to utilize an MLP, e.g., based on the Hadamard product of edge-level source and destination node features:
```python
z = x[edge_index[0]] * x[edge_index[1]]
out = MLP(z)
```
The Hadamard product achieves that link predictions are undirected/symmetric. You can also use concatenation to make your decoder directional.

**Reply by sinaziaee** [reply]
Thank you Matthias. You're always very helpful.
